agent/cache-types/materialized_view.go

```diff
@@ -0,0 +1,419 @@
+	// are returned from subscriptions.
+	SubscribeBackoffMax = 60 * time.Second
+)
+
+// StreamingClient is the interface we need from the gRPC client stub. Separate
+// interface simplifies testing.
+type StreamingClient interface {
+	Subscribe(ctx context.Context, in *stream.SubscribeRequest, opts ...grpc.CallOption) (stream.Consul_SubscribeClient, error)
+}
+
+// MaterializedViewState is the interface used to manage they type-specific
+// materialized view logic.
+type MaterializedViewState interface {
+	// InitFilter is called once when the view is constructed if the subscription
+	// has a non-empty Filter argument. The implementor is expected to create a
+	// *bexpr.Filter and store it locally so it can be used to filter events
+	// and/or results. Ideally filtering should occur inside `Update` calls such
+	// that we don't store objects in the view state that are just filtered when
+	// the result is returned, however in some cases it might not be possible and
+	// the type may choose to store the whole view and only apply filtering in the
+	// Result method just before returning a result.
+	InitFilter(expression string) error
+
+	// Update is called when one or more events are received. The first call will
+	// include _all_ events in the initial snapshot which may be an empty set.
+	// Subsequent calls will contain one or more update events in the order they
+	// are received.
+	Update(events []*stream.Event) error
+
+	// Result returns the type-specific cache result based on the state. When no
+	// events have been delivered yet the result should be an empty value type
+	// suitable to return to clients in case there is an empty result on the
+	// servers. The index the materialized view represents is maintained separtely
+	// and passed in in case the return type needs an Index field populating. This
+	// allows implementations to not worry about maintaining indexes seen during
+	// Update.
+	Result(index uint64) (interface{}, error)
+}
+
+// StreamingCacheType is the interface a cache-type needs to implement to make
+// use of streaming as the transport for updates from the server.
+type StreamingCacheType interface {
+	NewMaterializedViewState() MaterializedViewState
+	StreamingClient() StreamingClient
+	Logger() *log.Logger
+}
+
+// MaterializedView is a partial view of the state on servers, maintained via
+// streaming subscriptions. It is specialized for different cache types by
+// providing a MaterializedViewState that encapsulates the logic to update the
+// state and format it as the correct result type.
+//
+// The MaterializedView object becomes the cache.Result.State for a streaming
+// cache type and manages the actual streaming RPC call to the servers behind
+// the scenes until the cache result is discarded when TTL expires.
+type MaterializedView struct {
+	// Properties above the lock are immutable after the view is constructed in
+	// MaterializedViewFromFetch and must not be modified.
+	typ        StreamingCacheType
```

Comment:
Is there a reason it is `typ` instead of `type`? 
It is _shorter_ but I feel like the addition of the extra letter could just help provide some clarity where I was like "what is a `typ`? Oh! "

Reply:
Yeah it's gross but `type` is a reserved word in Go (see up there on line 73) so you can't use it for a field or variable name. We use `typ` everywhere else in the `cache` and `cache-types` package since Mitchell wrote it.

This is a bit like the ruby custom of naming things `klass` since `class` is a reserved word.